### PR TITLE
Argument length and argument constant optimization

### PR DESCRIPTION
**Pull request by yullin-ms**
Optimize LdLen_A argument length and LdElemI_A for the argument[constants]

### CODE REVIEW

File: lib/Backend/Func.cpp [hunk 1]
```diff
@@ -301,8 +303,10 @@ Func::Codegen(JitArenaAllocator *alloc, JITTimeWorkItem * workItem,
     Js::ScriptContextProfiler *const codeGenProfiler, const bool isBackgroundJIT)
 {
     bool rejit;
+    int rejitCounter = 0;
     do
     {
+        Assert(rejitCounter < 25);
```
Comment (rajatd) [inline lib/Backend/Func.cpp]: >rejitCounter [](start = 15, length = 12)

I don't quite get this - are we saying that repeated rejits upto 25 are allowed?

File: lib/Backend/GlobOpt.cpp [hunk 2]
```diff
@@ -15609,16 +15665,23 @@ GlobOpt::TrackArgumentsObject()
...
 }
 
 void
-GlobOpt::CannotAllocateArgumentsObjectOnStack()
+GlobOpt::CannotAllocateArgumentsObjectOnStack(Func * curFunc)
 {
+    if (curFunc != nullptr && curFunc->hasArgLenAndConstOpt)
+    {
+        Assert(!curFunc->GetJITOutput()->GetOutputData()->disableStackArgOpt);
+        curFunc->GetJITOutput()->GetOutputData()->disableStackArgOpt = true;
```
Comment (rajatd) [inline lib/Backend/GlobOpt.cpp]: >disableStackArgOpt [](start = 50, length = 18)

Why do we have to disable the stack args opt as a whole? Can't we just disable argLenAndConstOpt (and then check for it being enabled in OptArgLenAndConst) ?

File: lib/Backend/GlobOpt.cpp [hunk 3]
```diff
@@ -13083,6 +13084,61 @@ GlobOpt::OptArraySrc(IR::Instr ** const instrRef, Value ** src1Val, Value ** src
...
+            instr->m_func->hasArgLenAndConstOpt = true;
+        };
+        switch(instr->m_opcode)
+        {
+            case Js::OpCode::LdLen_A:
+            {
+                if (CurrentBlockData()->IsArgumentsOpnd(src1))
+                {
+                    IR::AddrOpnd* newopnd = IR::AddrOpnd::New(Js::TaggedInt::ToVarUnchecked(instr->m_func->actualCount - 1), IR::AddrOpndKindConstantVar, instr->m_func);
+                    replaceInstr(instr, newopnd, src1Val);
```
Comment (rajatd) [inline lib/Backend/GlobOpt.cpp]: >replaceInstr [](start = 20, length = 12)

For LdLen_A, does the Ld_A eventually get type specialized to Ld_I4 (int 32 type spec)?

Comment (yullin-ms) [reply lib/Backend/GlobOpt.cpp]: Yes

File: lib/Backend/Func.h [hunk 4]
```diff
@@ -715,12 +726,15 @@ static const unsigned __int64 c_debugFillPattern8 = 0xcececececececece;
     StackSym *          tempSymBool;
     uint32              loopCount;
     Js::ProfileId       callSiteIdInParentFunc;
+    InlineeFrameInfo*   argOptFrameInfo;
```
[the remark this answers is not in the document]
Comment (yullin-ms) [reply lib/Backend/Func.h]: ee1f312535ee71dbd82ce297c62641c5a331669f

File: lib/Backend/GlobOpt.cpp [hunk 5]
```diff
@@ -2442,6 +2442,19 @@ GlobOpt::OptInstr(IR::Instr *&instr, bool* isInstrRemoved)
...
+    if (instr->m_opcode == Js::OpCode::InlineeEnd &&
+        instr->m_func->hasArgLenAndConstOpt &&
+        instr->m_func->hasOptimizedAllArgConst)
+    {
+        instr->m_func->hasUnoptimizedArgumentsAccess = false;
+        if (DoInlineArgsOpt(instr->m_func))
+        {
+            instr->m_func->m_hasInlineArgsOpt = true;
+            instr->m_func->frameInfo = instr->m_func->argOptFrameInfo;
+        }
```
[the remark this answers is not in the document]
Comment (yullin-ms) [reply lib/Backend/GlobOpt.cpp]: ee1f312535ee71dbd82ce297c62641c5a331669f

File: lib/Backend/GlobOpt.cpp [hunk 6]
```diff
@@ -1603,13 +1608,15 @@ GlobOpt::OptArguments(IR::Instr *instr)
                     if (builtinFunction == Js::BuiltinFunction::JavascriptFunction_Apply)
                     {
                         CurrentBlockData()->ClearArgumentsSym(src1->AsRegOpnd());
+                        instr->m_func->canOptimizeInlineeArg = true;
```
Comment (rajatd) [inline lib/Backend/GlobOpt.cpp]: Won't this set to true undo any possible sets to false of `canOptimizeInlineeArg`? For example, if there was a TypeOfElem before ArgOut_A_InlineBuiltIn

**Comment by yullin-ms** [comment]
@dotnet-bot test OSX static_osx_osx_debug

**Comment by yullin-ms** [comment]
@dotnet-bot test OSX

**Comment by yullin-ms** [comment]
@dotnet-bot test all

**Comment by yullin-ms** [comment]
@dotnet-bot test OSX static_osx_osx_tes

**Comment by yullin-ms** [comment]
@dotnet-bot test OSX static_osx_osx_test, test OSX static_osx_osx_release, test OSX _no_jit_shared_osx_osx_test

**Pull request merged**